### PR TITLE
Garbage collect containers for deleted pods

### PR DESCRIPTION
**Pull request by timstclair**
Fixes https://github.com/kubernetes/kubernetes/issues/8347

/cc @dchen1107

### CODE REVIEW

**Comment by yujuhong** [comment]
Looks good overall with one question: Should we pass podManager in container/container_gc.go? It seems to me that every container runtime will need access to this for GC.

**Comment by timstclair** [comment]
Hmm, I think we could abstract out even more common logic from `dockertools/container_gc.go`, but since this is a 1.2 feature, how about I move forward with what I have and then follow up with the refactoring?

**Comment by timstclair** [comment]
@k8s-bot test node e2e experimental

**Comment by timstclair** [comment]
@k8s-bot test node e2e experimental

**Comment by timstclair** [comment]
Thanks @yujuhong - squashed & reapplying LGTM

**Comment by timstclair** [comment]
Removing LGTM to fix the e2e test.

**Comment by timstclair** [comment]
Verified tests pass locally, but they currently fail in the e2e_node because of docker permissions. In the interest of getting this in for 1.2, I'm disabling the test for the time being.

@yujuhong PTAL

/cc @pwittrock FYI

**Comment by yujuhong** [comment]
Sorry I missed this PR somehow. Will review today.

File: test/e2e_node/garbage_collector_test.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+		)
+
+		containers, err := dockerClient.ListContainers(docker.ListContainersOptions{All: true})
+		Expect(err).To(BeNil(), fmt.Sprintf("Error listing containers %v", err))
+		initialContainerCount := len(containers)
+
+		// Start pods.
+		podNames := make([]string, numPods)
+		podContainers := []api.Container{getPauseContainer()}
+		for i := 0; i < numPods; i++ {
```
Comment (yujuhong) [inline test/e2e_node/garbage_collector_test.go]: nit: this (creating N pods) can be moved to util to increase code reuse.

File: test/e2e_node/garbage_collector_test.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+	})
+
+	It("Should garbage collect deleted pods", func() {
+		Skip("Requires docker permissions") // FIXME
+
+		const (
+			// The acceptable delta when counting containers.
+			epsilon = 5
+			// The number of pods to create & delete.
+			numPods = 90
```
Comment (yujuhong) [inline test/e2e_node/garbage_collector_test.go]: Do we need 90 pods?

File: test/e2e_node/garbage_collector_test.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+		test: func(x int) bool { return x >= val },
+	}
+}
+
+// Wait for at least count containers to be running if atleast is true, or at most count containers
+// to be running if atleast is false.
+func waitForContainerCount(dockerClient *docker.Client, cond condition) error {
+	const (
+		pollPeriod  = 10 * time.Second
+		pollTimeout = 5 * time.Minute
```
Comment (yujuhong) [inline test/e2e_node/garbage_collector_test.go]: nit: add a comment on the GC period

Comment (yujuhong) [inline test/e2e_node/garbage_collector_test.go]: Should probably add the comment in line 74

File: test/e2e_node/garbage_collector_test.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+		// Start pods.
+		podNames := make([]string, numPods)
+		podContainers := []api.Container{getPauseContainer()}
+		for i := 0; i < numPods; i++ {
+			podNames[i] = fmt.Sprintf("pod-%d", i)
+			createPod(cl, podNames[i], podContainers, nil)
+		}
+
+		// Wait for containers to start.
+		Expect(waitForContainerCount(dockerClient, atLeast(numPods))).To(BeNil())
```
Comment (yujuhong) [inline test/e2e_node/garbage_collector_test.go]: shouldn't the total number of containers to be 2* numPods? (infra+user)

**Comment by yujuhong** [comment]
Looks good overall with some nits/questions.

I am also ok with checking in the change without the test for the time being, if the test can'r really be run at this point.

**Comment by yujuhong** [comment]
We should also update `garbage-collection.md` with this new behavior.

File: test/e2e_node/garbage_collector_test.go (hunk 1 above)
Comment (yujuhong) [inline test/e2e_node/garbage_collector_test.go]: I think we can reuse https://github.com/kubernetes/kubernetes/blob/5b415a7aede5338595b06e4f4625736b18c185e3/test/e2e/util.go#L1940

**Comment by timstclair** [comment]
Pulled out e2e_node test (will send as a separate PR), and updated `garbage-collection.md`. PTAL

**Comment by yujuhong** [comment]
Build error:
```
_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/dockertools/fake_manager.go:49: undefined: pod.NewFakeMirrorClient
```

**Comment by timstclair** [comment]
Fixed the build error. Unfortunately it's a little messy since I needed to avoid pulling in pod/testing for the fake docker.

**Comment by timstclair** [comment]
Squashed.

**Comment by timstclair** [comment]
Adding P1 since this is a 1.2 feature.

**Comment by eparis** [comment]
ok to test

**Comment by ixdy** [comment]
@k8s-bot test this, issue #22435

**Pull request merged**